### PR TITLE
Bigtable: 06. Surface - CheckAndMutate

### PR DESCRIPTION
**Pull request by igorbernstein2**
This has been reviewed @kevinsi4508. This is ready for @pongad & @garrettjonesgoogle

### CODE REVIEW

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/ConditionalRowMutation.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  private final String tableId;
+  private final CheckAndMutateRowRequest.Builder builder = CheckAndMutateRowRequest.newBuilder();
+
+  private ConditionalRowMutation(String tableId, ByteString rowKey) {
+    this.tableId = tableId;
+    builder.setRowKey(rowKey);
+  }
+
+  /** Creates a new instance of the mutation builder. */
+  public static ConditionalRowMutation create(String tableId, String key) {
```
Comment (kevinsi4508) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/ConditionalRowMutation.java]: "key" or "rowKey"? be consistent.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java [hunk 2]
```diff
@@ -338,4 +339,46 @@ public void readRowsAsync(Query query, ResponseObserver<Row> observer) {
...
   }
+
+  /**
+   * Convenience method to asynchronously mutate a row atomically based on the output of a filter.
+   *
+   * <p>Sample code:
+   *
+   * <pre>{@code
+   * InstanceName instanceName = InstanceName.of("[PROJECT]", "[INSTANCE]");
+   * try (BigtableClient bigtableClient = BigtableClient.create(instanceName)) {
```
Comment (kevinsi4508) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java]: An example with condition() is better I believe.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/ConditionalRowMutation.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+   * results are yielded, either the mutations added via {@link #then(Mutation)} or {@link
+   * #otherwise(Mutation)} will be executed. If unset, checks that the row contains any values at
+   * all.
+   */
+  public ConditionalRowMutation condition(@Nonnull Filter condition) {
+    Preconditions.checkNotNull(condition);
+    Preconditions.checkState(
+        !builder.hasPredicateFilter(),
+        "Can only have a single condition, please use a Filters#chain or Filters#interleave filter instead");
+    //TODO: verify that the condition does not use any FILTERS.condition() filters
```
Comment (kevinsi4508) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/ConditionalRowMutation.java]: Could you add this to the method doc?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/ConditionalRowMutation.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+    builder.setPredicateFilter(condition.toProto());
+
+    return this;
+  }
+
+  /**
+   * Adds changes to be atomically applied to the specified row if the condition yields at least one
```
Comment (kevinsi4508) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/ConditionalRowMutation.java]: Could you make the comment here a bit more clear?
"changes" are the same as "mutations"?  What are "Entries" then? mutations?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/Mutation.java [hunk 5]
```diff
@@ -28,7 +28,8 @@
 
 /**
  * The concrete implementation of {@link MutationApi} that can be used to create and represent a
- * list of mutations. This class is meant to used as child of other classes.
+ * list of mutations. This class is meant to used as child of other classes (like {@link
```
Comment (kevinsi4508) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/Mutation.java]: "meant to be used"?
It is not clear to me what is "child of other classes" means here.

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/models/ConditionalRowMutationTest.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+import com.google.bigtable.v2.TableName;
+import com.google.cloud.bigtable.data.v2.internal.RequestContext;
+import com.google.common.truth.Truth;
+import com.google.protobuf.ByteString;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class ConditionalRowMutationTest {
```
Comment (kevinsi4508) [inline google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/models/ConditionalRowMutationTest.java]: nit: test toProto() fails when getTrueMutationsList() && getFalseMutationsList() are empty?

**Comment by igorbernstein2** [comment]
All feedback should be addressed, PTAL

**Comment by kevinsi4508** [comment]
LGTM

**Comment by igorbernstein2** [comment]
@pongad & @garrettjonesgoogle can you take a look?

**Comment by igorbernstein2** [comment]
Please merge when ready

**Pull request merged**